### PR TITLE
Added test cases for Pallet Protos and Pallet Tickets

### PR DESCRIPTION
**Pull request by aang114**

### CODE REVIEW

**Comment by aang114** [comment]
@sinkingsugar let me know if I missed any edge case for the test cases. Thanks

File: .github/workflows/check.yml [hunk 1]
```diff
@@ -67,7 +67,7 @@ jobs:
 
       - name: Check Test
         run: |
-          cargo test
+          cargo test --no-fail-fast
```
Comment (sinkingsugar) [inline .github/workflows/check.yml:70]: what does this to? reports error on first failure?

File: pallets/protos/Cargo.toml [hunk 2]
```diff
@@ -100,6 +100,14 @@ serde_json = { version = '1.0.79', default-features = false, features = ['alloc'
 protos = { version = "0.1.11", default-features = false }
 base58 = { version = "0.2.0", default-features = false }
 
+# [dev-dependencies]
+# sp-clamor = { version = '0.1.0', path = '../../primitives/clamor'}
+# pallet-detach = {path = '../detach'}
+# pallet-frag = {path = '../frag'}
+
+[dev-dependencies]
+pallet-tickets = {path = '../tickets', default-features = false, features=['compile-dummy-data']}
```
Comment (sinkingsugar) [inline pallets/protos/Cargo.toml]: this is already a normal dependency no? should be.

File: pallets/protos/src/dummy_data.rs [hunk 3]
```diff
@@ -0,0 +1,203 @@
+use crate::*;
+
+pub use pallet_tickets::dummy_data::{
+	Lock, Link,
+};
+
+
+use sp_core::{
+
```
Comment (sinkingsugar) [inline pallets/protos/src/dummy_data.rs]: remove empty space

File: pallets/protos/src/dummy_data.rs [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+use sp_clamor::{
+	Hash256, CID_PREFIX,
+};
+
+use protos::categories::{
+	Categories, TextCategories,
+};
+
+
```
Comment (sinkingsugar) [inline pallets/protos/src/dummy_data.rs:17]: remove empty space

File: pallets/protos/src/dummy_data.rs [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+		compute_data_hash(&self.data)
+	}
+
+	pub fn get_proto_cid(&self) -> Vec<u8> {
+		compute_data_cid(&self.data)
+	}
+}
+
+
+
```
Comment (sinkingsugar) [inline pallets/protos/src/dummy_data.rs]: remove empty spaces etc etc

File: pallets/protos/src/dummy_data.rs [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+/// NOTE: All `ProtoFragment`-type fields found in `DummyData` have no references
+pub struct DummyData {
+	pub proto_fragment: ProtoFragment,
+	pub proto_fragment_second: ProtoFragment,
+	pub patch: Patch,
+	pub metadata: Metadata,
+	// This is a stake that will succeed
+	pub stake: Stake,
+	
+
```
Comment (sinkingsugar) [inline pallets/protos/src/dummy_data.rs]: 😭

File: pallets/protos/src/dummy_data.rs [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+	pub account_id_second: sp_core::ed25519::Public,
+
+	pub ethereum_account_id: H160,
+	 
+}
+
+impl DummyData {
+    pub fn new() -> Self {
+
```
Comment (sinkingsugar) [inline pallets/protos/src/dummy_data.rs]: 😭

**Review by sinkingsugar — COMMENTED** [review]
Please follow the current formatting style (keep code compact)
Also use `cargo fmt` often.
Will resume review later.

File: .github/workflows/check.yml (hunk 1 above)
Comment (alebaffa) [reply .github/workflows/check.yml:70]: `--no-fail-fast: Run all tests regardless of failure. Without this flag, Cargo will exit after the first executable fails.`
The CI at the end will fail anyway in case of any failed tests, but it will also run all the other tests too instead of stopping immediately. I guess you put this to avoid wasting time discovering new failing tests one after the other, right?

Comment (aang114) [reply .github/workflows/check.yml:70]: Yup, it's so that all tests run so I avoid wasting time

File: pallets/protos/Cargo.toml (hunk 2 above)
Comment (aang114) [reply pallets/protos/Cargo.toml]: Yes. But the normal dependency doesn't enable the feature "compile-dummy-data"

(Dev-Dependencies only run when doing `cargo test` btw)

File: pallets/protos/src/dummy_data.rs [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+pub struct Metadata {
+	pub proto_fragment: ProtoFragment,
+	pub metadata_key: Vec<u8>,
+	pub data: Vec<u8>,
+}
+impl Metadata {
+	pub fn get_data_hash(&self) -> Hash256 {
+		compute_data_hash(&self.data)
+	}
+	pub fn get_data_cid(&self) -> Vec<u8> {
```
Comment (alebaffa) [inline pallets/protos/src/dummy_data.rs]: This function is never used.

File: pallets/protos/src/tests.rs [hunk 9]
```diff
@@ -1,381 +1,920 @@
...
+				tickets_info: TicketsInfo::default(),
+			};
+
+			// Ensure that this test case fails if a new field is ever added to the `Proto` struct
+			match proto_struct {
+				s if s == correct_proto_struct => (),
+				_ => panic!("The correct `Proto` struct was not saved in the StorageMap `Protos`"),
+			}
+
+
```
Comment (alebaffa) [inline pallets/protos/src/tests.rs]: remove empty space.

File: pallets/protos/src/tests.rs [hunk 10]
```diff
@@ -1,381 +1,920 @@
...
+	#[test]
+	fn upload_should_work_if_user_staked_enough() {
+
+		new_test_ext().execute_with(|| {
+			
+			let dd = DummyData::new();
+
+			let stake = dd.stake;
+	
+			upload(dd.account_id, &stake.proto_fragment);
```
Comment (alebaffa) [inline pallets/protos/src/tests.rs]: All these functions
```rust
upload(....)
lock(....)
link(...)
stake(....)
```
return a Result that should be handled. It is not a problem, I mean, it's just that they create annoying warnings in the compiler.
A quick fix could be to ignore the returned value:
```rust
_ = upload(....)
_ = lock(....)
_ = link(...)
_ = stake(....)
```

File: pallets/tickets/src/dummy_data.rs [hunk 11]
```diff
@@ -0,0 +1,263 @@
+use crate::*;
```
Comment (alebaffa) [inline pallets/tickets/src/dummy_data.rs]: This is unused import. Can be removed.

File: pallets/tickets/src/dummy_data.rs [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+
+use codec::{Decode, Encode};
```
Comment (alebaffa) [inline pallets/tickets/src/dummy_data.rs]: Decode is unused import, can be removed.

File: pallets/tickets/src/tests.rs [hunk 13]
```diff
@@ -1 +1,600 @@
...
+        TypedGet,
+    }
+};
+
+use frame_support::dispatch::DispatchResult;
+
+use sp_core::{
+
+    H160,
+
```
Comment (alebaffa) [inline pallets/tickets/src/tests.rs]: lots of empty spaces in this file that can be removed.

File: pallets/tickets/src/tests.rs [hunk 14]
```diff
@@ -1 +1,600 @@
...
+        new_test_ext().execute_with(|| {
+
+			let dd = DummyData::new();
+
+            let link = dd.link;
+
+
+            assert_ok!(
+                link_(link.clamor_account_id, &link.get_link_signature())
```
Comment (alebaffa) [inline pallets/tickets/src/tests.rs]: Replace 
```rust
link_(...)
```
with
```rust
_ = link_(...)
```
to fix the annoying warning from compiler due to unhandled returned value.

Comment (alebaffa) [reply pallets/tickets/src/tests.rs]: @aang114 generally speaking, if you use rust-analyzer in VSCode you will see immediately all the warnings. Very handy.

Comment (aang114) [reply pallets/tickets/src/tests.rs]: Oh you were asking about rust-analyzer earlier because of this 😅 . 

Yeah I use it, but I never considered warnings important 😅. Will be mindful of them from now on

Comment (alebaffa) [reply pallets/tickets/src/tests.rs]: Yeah 👍 it's technical debt anyway. Warnings are not important until they are, in my experience 😄 better to keep only the very necessary ones.

Comment (sinkingsugar) [reply pallets/tickets/src/tests.rs]: @aang114 warnings are red flags, get rid of them ASAP.

File: pallets/tickets/src/tests.rs (hunk 13 above)
Comment (sinkingsugar) [reply pallets/tickets/src/tests.rs]: 👍👍👍

File: pallets/protos/src/tests.rs (hunk 10 above)
Comment (sinkingsugar) [reply pallets/protos/src/tests.rs]: Shouldn't the result be handled though? At least `unwrap`.

File: pallets/protos/src/tests.rs (hunk 9 above)
Comment (sinkingsugar) [reply pallets/protos/src/tests.rs]: 👍👍👍

File: pallets/accounts/Cargo.toml [hunk 15]
```diff
@@ -103,6 +103,11 @@ serde = { version = "1.0.136", features = ["derive"], optional = true }
 serde_json = { version = '1.0.79', default-features = false, features = ['alloc'] }
 ethabi = { version = "17.0.0", default-features = false }
 
+parking_lot = "0.12.1" # copied from https://github.com/JoshOrndorff/recipes/blob/master/pallets/ocw-demo/Cargo.toml#L11
+libsecp256k1 = {version = "0.7.0", default-features = false} # https://crates.io/crates/libsecp256k1
+
```
Comment (alebaffa) [inline pallets/accounts/Cargo.toml:113]: Please remove these spaces here.

File: pallets/accounts/Cargo.toml [hunk 16]
```diff
@@ -124,3 +129,8 @@ std = [
 	'sp-core/std',
 ]
 try-runtime = ['frame-support/try-runtime']
+
+compile-dummy-data=[]
+
+
```
Comment (alebaffa) [inline pallets/accounts/Cargo.toml]: Please remove these spaces here.

File: pallets/accounts/src/mock.rs [hunk 17]
```diff
@@ -1,19 +1,57 @@
...
+
+
+use parking_lot::RwLock;
+
+use sp_core::{
+
+	H256,
+
+	ed25519::Signature,
+
```
Comment (alebaffa) [inline pallets/accounts/src/mock.rs]: A bit messy here with spaces.

File: pallets/accounts/src/mock.rs [hunk 18]
```diff
@@ -1,19 +1,57 @@
...
+use sp_keystore::{
+    testing::KeyStore,
+    KeystoreExt,
+    SyncCryptoStore,
+};
+
+use std::sync::Arc;
+
+
```
Comment (alebaffa) [inline pallets/accounts/src/mock.rs]: Too many blank lines here.

File: pallets/accounts/src/mock.rs [hunk 19]
```diff
@@ -138,17 +192,59 @@ impl pallet_proxy::Config for Test {
...
+pub fn new_test_ext_with_ocw() -> (
+	sp_io::TestExternalities, 
+	Arc<RwLock<PoolState>>, 
+	Arc<RwLock<OffchainState>>, 
+	sp_core::ed25519::Public
+) {
+	const PHRASE: &str =
+		"news slush supreme milk chapter athlete soap sausage put clutch what kitten";
+
+
```
Comment (alebaffa) [inline pallets/accounts/src/mock.rs]: Too many blank lines here too.

File: pallets/accounts/src/dummy_data.rs [hunk 20]
```diff
@@ -0,0 +1,240 @@
...
+	let hashed_message = keccak_256(
+		&[b"\x19Ethereum Signed Message:\n32", &hashed_message[..]].concat()
+	);
+	
+	ethereum_account_pair.sign_prehashed(&hashed_message)
+}
+
+pub fn get_ethereum_account_id_from_ecdsa_public_struct(ecdsa_public_struct: &ecdsa::Public) -> H160 {
+
+		let compressed_public_key = ecdsa_public_struct.0;
```
Comment (alebaffa) [inline pallets/accounts/src/dummy_data.rs]: The code in this function seems to have a different alignment compared to the other ones.
There are other functions like this too.

File: pallets/accounts/src/mock.rs [hunk 21]
```diff
@@ -1,19 +1,57 @@
...
+};
+
+use std::sync::Arc;
+
+
+
 use sp_runtime::{
 	testing::{Header, TestXt},
 	traits::{BlakeTwo256, Extrinsic as ExtrinsicT, IdentifyAccount, IdentityLookup, Verify},
+
```
Comment (alebaffa) [inline pallets/accounts/src/mock.rs]: blank line to remove.

**Comment by aang114** [comment]
Hi guys (@sinkingsugar, @alebaffa), please let me know if the test cases I have written are sufficient. Thanks :)

I have a feeling I might not have covered all the test cases for `link()` (see `pallets/accounts/tests.rs`)

File: pallets/accounts/src/tests.rs [hunk 22]
```diff
@@ -1 +1,550 @@
...
+		});
+	}
+
+	#[test]
+	// #[ignore = "seems there's a bug with linking"]
+	fn link_should_not_work_if_origin_parameter_is_invalid() {
+		new_test_ext().execute_with(|| {
+			let dd = DummyData::new();
+
+			let link = Link { clamor_account_id: dd.account_id, ..dd.link };
```
Comment (alebaffa) [inline pallets/accounts/src/tests.rs]: what does it mean `..dd.link` here? Why do you need `..` ?

File: pallets/accounts/src/tests.rs [hunk 23]
```diff
@@ -1 +1,550 @@
...
+	}
+
+	#[test]
+	fn link_should_not_work_if_signature_parameter_is_invalid() {
+		new_test_ext().execute_with(|| {
+			let dd = DummyData::new();
+
+			let link = Link { link_signature: dd.link_signature, ..dd.link };
+
+			dd.link;
```
Comment (alebaffa) [inline pallets/accounts/src/tests.rs]: What does this do?

File: pallets/accounts/src/tests.rs [hunk 24]
```diff
@@ -1 +1,550 @@
...
+		});
+	}
+}
+
+mod sync_frag_locks_tests {
+	use super::*;
+
+	#[test]
+	fn sync_frag_locks_should_work() {
+		let (mut t, pool_state, offchain_state, ed25519_public_key) = new_test_ext_with_ocw();
```
Comment (alebaffa) [inline pallets/accounts/src/tests.rs]: What does `ocw` mean?

File: pallets/accounts/src/tests.rs [hunk 25]
```diff
@@ -1 +1,550 @@
...
+
+		sp_clamor::init(geth_url);
+
+		let latest_block_number = lock
+			.data
+			.block_number // ensure that `lock.block_number` exists by making `latest_block_number` greater than or equal to it
+			.saturating_add(<Test as pallet_accounts::Config>::EthConfirmations::get())
+			.saturating_add(69);
+
+		offchain_state.write().expect_request(testing::PendingRequest {
```
Comment (alebaffa) [inline pallets/accounts/src/tests.rs]: I think it would be better to extract all this "plumbing" with offchain_state to an external function and keep only the asserts into the tests. All these pieces of code make the tests pretty unreadable.

File: pallets/accounts/src/tests.rs (hunk 22 above)
Comment (aang114) [reply pallets/accounts/src/tests.rs]: `..dd.link` allows us to use the same fields as `dd.link` for the remaining fields  (see: https://doc.rust-lang.org/reference/expressions/struct-expr.html#functional-update-syntax).

File: pallets/accounts/src/tests.rs (hunk 24 above)
Comment (sinkingsugar) [reply pallets/accounts/src/tests.rs]: off chain worker

File: pallets/protos/src/tests.rs [hunk 26]
```diff
@@ -1,385 +1,851 @@
...
+		new_test_ext().execute_with(|| {
+			let dd = DummyData::new();
+
+			let stake = dd.stake;
 
-#[test]
-fn patch_should_not_work_if_detached() {
-	let keystore = KeyStore::new();
-	let mut t = new_test_ext();
+			_ = upload(dd.account_id, &stake.proto_fragment);
```
Comment (sinkingsugar) [inline pallets/protos/src/tests.rs]: we are ignoring a lot of results here...

Comment (aang114) [reply pallets/protos/src/tests.rs]: Since `upload()` is tested. We can safely assume that if all the `upload()` tests worked, then this should also return `Ok(())`?

File: pallets/accounts/src/tests.rs (hunk 23 above)
Comment (aang114) [reply pallets/accounts/src/tests.rs]: typo, removing

File: pallets/accounts/src/tests.rs (hunk 25 above)
Comment (aang114) [reply pallets/accounts/src/tests.rs]: done

File: pallets/protos/src/tests.rs (hunk 26 above)
Comment (alebaffa) [reply pallets/protos/src/tests.rs]: I wouldn't do it because this can create situations when the `upload` function is broken, but these tests still would work, making the result of the test suite not clear in the CI. It's better to have all tests failing, regardless of where they are, when a function is not working.

Comment (aang114) [reply pallets/protos/src/tests.rs]: makes sense. I'll do `unwrap` instead of `assert_ok!`.

othherwise, there'll be a lot of `assert_ok`s 😆

Comment (aang114) [reply pallets/protos/src/tests.rs]: so we won't know what we're testing

**Comment by aang114** [comment]
> Better to use a `assert_ok`, yes. Panic should never occur.

I have changed it to `assert_ok`.

Btw do let me know if the test cases I have written are sufficient. I have a feeling I might not have covered all the test cases for `link()` (see pallets/accounts/tests.rs). Thanks

**Comment by aang114** [comment]
Hi guys, could you review this PR soon (hopefully today) - since it was supposed to be done 2 SPRINTs (and only after this is merged can I create the PR for the test cases of pallet fragments). Sorry about this 😅

**Comment by alebaffa** [comment]
The function `get_protos` has no tests. But we can do this in another PR later.

**Review by alebaffa — APPROVED** [review]

File: pallets/accounts/src/mock.rs [hunk 27]
```diff
@@ -22,7 +38,7 @@ frame_support::construct_runtime!(
 		UncheckedExtrinsic = UncheckedExtrinsic,
 	{
 		System: frame_system::{Pallet, Call, Config, Storage, Event<T>},
-		FragPallet: pallet_frag::{Pallet, Call, Storage, Event<T>},
+		AccountsPallet: pallet_accounts::{Pallet, Call, Storage, Event<T>, ValidateUnsigned},
```
Comment (sinkingsugar) [inline pallets/accounts/src/mock.rs]: remove xPallet suffix, it's not used for other

File: pallets/accounts/src/mock.rs [hunk 28]
```diff
@@ -138,17 +170,52 @@ impl pallet_proxy::Config for Test {
...
+	let mut ext = sp_io::TestExternalities::new(t);
+
+	ext.execute_with(|| System::set_block_number(1)); // if we don't execute this line, Events are not emitted from extrinsics (I don't know why this is the case though)
+
+	ext
+}
+
+pub fn new_test_ext_with_ocw() -> (
+	sp_io::TestExternalities,
+	Arc<RwLock<PoolState>>,
```
Comment (sinkingsugar) [inline pallets/accounts/src/mock.rs:195]: oh wow, where did you find this out? nice

Comment (aang114) [reply pallets/accounts/src/mock.rs:195]: https://github.com/paritytech/substrate/blob/master/frame/examples/offchain-worker/src/tests.rs#L208

File: pallets/protos/src/mock.rs [hunk 29]
```diff
@@ -44,7 +35,7 @@ frame_support::construct_runtime!(
 		DetachPallet: pallet_detach::{Pallet, Call, Storage, Event<T>},
 		CollectiveFlip: pallet_randomness_collective_flip::{Pallet, Storage},
 		Balances: pallet_balances::{Pallet, Call, Storage, Config<T>, Event<T>},
-		Accounts: pallet_accounts::{Pallet, Call, Storage, Event<T>},
+		AccountsPallet: pallet_accounts::{Pallet, Call, Storage, Event<T>},
```
Comment (sinkingsugar) [inline pallets/protos/src/mock.rs]: Accounts

File: pallets/protos/src/tests.rs [hunk 30]
```diff
@@ -1,385 +1,885 @@
...
+				mock::Event::from(pallet_protos::Event::Uploaded {
+					proto_hash: proto.get_proto_hash(),
+					cid: proto.get_proto_cid()
+				})
+			);
+		});
+	}
+
+	#[test]
+	fn upload_should_not_works_if_proto_hash_exists() {
```
Comment (sinkingsugar) [inline pallets/protos/src/tests.rs]: Btw can you fix those names too? With proper English. The previous engineer didn't really compute... "works" 😝

File: pallets/protos/src/tests.rs [hunk 31]
```diff
@@ -1,385 +1,885 @@
...
+		signer: <Test as frame_system::Config>::AccountId,
+		proto: &ProtoFragment,
+		stake_amount: &u64,
+	) -> DispatchResult {
+		ProtosPallet::stake(Origin::signed(signer), proto.get_proto_hash(), stake_amount.clone())
+	}
+
+	// TODO
+	#[test]
+	#[ignore]
```
Comment (sinkingsugar) [inline pallets/protos/src/tests.rs:621]: Nice 👍

**Review by sinkingsugar — CHANGES_REQUESTED** [review]

File: pallets/accounts/src/mock.rs (hunk 27 above)
Comment (aang114) [reply pallets/accounts/src/mock.rs]: done for everything except `Protos`, since there also exists a StorageMap with the name `Protos`

**Review by sinkingsugar — APPROVED** [review]
